### PR TITLE
ACS-2245: Mitigate intermittent failure of FixedAclUpdaterTest

### PR DESCRIPTION
**Pull request by dsibilio**
Introducing a JUnit Rule to mitigate the flakiness of specific test methods which are deemed to be valuable assets.

The solution is inspired from the `maven-surefire-plugin` [rerunFailingTestsCount configuration option](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html), but has the advantage of being applicable in a selective, configurable, and granular way, therefore not hiding potentially dangerous/unneeded tests' flakiness but only coming into play when deemed necessary.

#### Before Mitigation

| Runs  | Failures |
| ------ | -------- |
| 10      | 2           |

#### After Mitigation

| Runs  | Failures |
| ------ | -------- |
| 18      | 0           |

### CODE REVIEW

File: repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import org.junit.runners.model.Statement;
+
+/**
+ * This JUnit rule can be used to turn existing test code into repeatable tests.
+ * The test methods marked with the {@link RepeatAtMost} annotation will be repeated at most the specified
+ * amount of times, stopping at the first successful execution.
+ *
+ * @author Domenico Sibilio
+ */
+public class RepeatAtMostRule implements TestRule
```
Comment (jamalkm) [inline repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java:48]: I'd suggest to rename `Repeat...` to `Retry...` (i.e. `RetryAtMostRule`) as the execution stops at the first successful run.

File: repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+        if (repeatAtMost != null)
+        {
+            return new RepeatAtMostTestStatement(statement, description, repeatAtMost.value());
+        }
+
+        return statement;
+    }
+
+    private static class RepeatAtMostTestStatement extends Statement
```
Comment (jamalkm) [inline repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java:65]: Same as the above (class name Repeat... -> Retry...).

File: repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        private final Statement statement;
+        private final Description description;
+        private final int times;
```
Comment (jamalkm) [inline repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java:69]: I believe renaming `times` to `retryCount` makes it clearer.

File: repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    /**
+     * This annotation is a marker used to identify a JUnit &#64;{@link Test} method as a repeatable test.
+     */
+    @Target(ElementType.METHOD)
+    @Retention(RetentionPolicy.RUNTIME)
+    @Documented
+    public @interface RepeatAtMost
```
Comment (jamalkm) [inline repository/src/test/java/org/alfresco/util/test/junitrules/RepeatAtMostRule.java:125]: Same as others (Repeat -> Retry)

**Review by jamalkm — COMMENTED** [review]
Nice work. I really liked the solution and the clean implementation + test. 
However, I just left a couple of very minor comments to do with class/variable names.

**Review by damianujma — APPROVED** [review]
LGTM.

**Review by jamalkm — APPROVED** [review]
LGTM

**Review by MateuszBiel — APPROVED** [review]

**Pull request merged**